File: Graph/542. 01 Matrix/updateMatrix.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
 var updateMatrix = function(mat) {
    let n = mat.length; 
    let m = mat[0].length; 

    // build columns
    let distances = new Array(n);
    // traverse to n
    for (let i = 0; i < n; i++) {
        distances[i] = new Array(m).fill(0);
    }
    
    let longestDistance = n - 1 + m - 1;
    
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if(mat[i][j] !== 0) {
                let up = (i > 0) ? distances[i - 1][j] : longestDistance;
                let left = (j > 0) ? distances[i][j - 1] : longestDistance;
                distances[i][j] = Math.min(up, left) + 1;
            }
        }
    }
    
    for (let i = n - 1; i >= 0; i--) {
        for (let j = m - 1; j >= 0; j--) {
            if(mat[i][j] !== 0) {
                let down = (i < n - 1) ? distances[i + 1][j] : longestDistance;
                let right = (j < m - 1) ? distances[i][j + 1] : longestDistance;
                distances[i][j] = Math.min(Math.min(down, right) + 1, distances[i][j]);
            }
        }
    }
     return distances;
};

File: Graph/542. 01 Matrix/updateMatrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        n = len(mat)
        m = len(mat[0])
        distances = [[0 for _ in range(m)] for _ in range(n)]
        
        longest_distance = m - 1 + n - 1
        
        for i in range(n):
            for j in range(m):
                if mat[i][j] != 0:
                    up = distances[i - 1][j] if (i > 0) else longest_distance
                    left = distances[i][j - 1] if (j > 0) else longest_distance
                    distances[i][j] = min(up, left) + 1
        
        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                if mat[i][j] != 0:
                    down = distances[i + 1][j] if (i < n - 1) else longest_distance
                    right = distances[i][j + 1] if (j < m - 1) else longest_distance
                    distances[i][j] = min(min(down, right) + 1, distances[i][j])
                    
        return distances